bgpd/bgpd.c

```diff
@@ -6587,6 +6588,173 @@ int peer_unsuppress_map_unset(struct peer *peer, afi_t afi, safi_t safi)
 	return 0;
 }
 
+static void peer_update_rmap_filter_data(struct peer *peer, afi_t afi,
+					 safi_t safi, const char *rmap_name1,
+					 struct route_map *rmap1,
+					 const char *rmap_name2,
+					 struct route_map *rmap2,
+					 uint8_t config_flags)
+{
+	struct bgp_filter *filter;
+	bool filter_exists = false;
+
+	filter = &peer->filter[afi][safi];
+
+	if (CHECK_FLAG(config_flags, BGP_PEER_ADVERTISE_MAP)) {
+		/* advertise-map is already configured. */
+		if (filter->advmap.aname) {
+			filter_exists = true;
+			XFREE(MTYPE_BGP_FILTER_NAME, filter->advmap.aname);
+			XFREE(MTYPE_BGP_FILTER_NAME, filter->advmap.cname);
+		}
+		route_map_counter_decrement(filter->advmap.amap);
+
+		/* Removed advertise-map configuration */
+		if (!CHECK_FLAG(config_flags, BGP_PEER_RMAP_SET)) {
+			memset(filter, 0, sizeof(struct bgp_filter));
+
+			/* decrement condition_filter_count delete timer if
+			 * this is the last advertise-map to be removed.
+			 */
+			if (filter_exists)
+				bgp_conditional_adv_disable(peer, afi, safi);
+
+			return;
+		}
+
+		/* Update filter data with newly configured values. */
+		filter->advmap.aname =
+			XSTRDUP(MTYPE_BGP_FILTER_NAME, rmap_name1);
+		filter->advmap.cname =
+			XSTRDUP(MTYPE_BGP_FILTER_NAME, rmap_name2);
+		filter->advmap.amap = rmap1;
+		filter->advmap.cmap = rmap2;
+		filter->advmap.condition =
+			CHECK_FLAG(config_flags, BGP_PEER_CONDITION_EXIST);
+		route_map_counter_increment(filter->advmap.amap);
+		peer->advmap_config_change[afi][safi] = true;
+
+		/* Increment condition_filter_count and/or create timer. */
+		if (!filter_exists) {
+			filter->advmap.advertise = ADVERTISE;
+			bgp_conditional_adv_enable(peer, afi, safi);
+		}
+	}
+}
+
+/* Set advertise-map to the peer but do not process peer route updates here.  *
+ * Hold filter changes until the conditional routes polling thread is called  *
+ * AS we need to advertise/withdraw prefixes (in advertise-map) based on the  *
+ * condition (exist-map/non-exist-map) and routes(specified in condition-map) *
+ * in BGP table. So do not call peer_on_policy_change() here, only create     *
+ * polling timer thread, update filters and increment condition_filter_count.
+ */
+int peer_advertise_map_set(struct peer *peer, afi_t afi, safi_t safi,
+			   const char *advertise_name,
+			   struct route_map *advertise_map,
+			   const char *condition_name,
+			   struct route_map *condition_map, bool condition)
+{
+	uint8_t config_flags = 0;
+	struct peer *member;
+	struct listnode *node, *nnode;
+
+	SET_FLAG(config_flags, BGP_PEER_RMAP_SET);
+	SET_FLAG(config_flags, BGP_PEER_ADVERTISE_MAP);
+	if (condition)
+		SET_FLAG(config_flags, BGP_PEER_CONDITION_EXIST);
```

Comment:
Can't this be reused with af_flags as well? My point is to reuse and keep using existing places for flags. PEER_FLAG_ADV_MAP_CONDITION_EXIST or so.

Reply:
These flags were introduced just to avoid multiple arguments to the function (which is defined to avoid duplicate code in set and unset functions) in case we want to use the same function for route-map and unsuppress-map.
Anyways, I will remove all these flags and make the function specific to advertise-map filter update.
i.e changed 
peer_update_rmap_filter_data(...., config_flags) to peer_advertise_map_filter_update(...., bool condition, bool set)